test/core/promise/activity_test.cc

```diff
@@ -321,6 +325,73 @@ TEST(WakerTest, CanWakeupEmptyWaker) {
   Waker().Wakeup();
 }
 
+TEST(AtomicWakerTest, CanWakeupEmptyWaker) {
+  // Empty wakers should not do anything upon wakeup.
+  AtomicWaker waker;
+  EXPECT_FALSE(waker.Armed());
+  waker.Wakeup();
+}
+
+TEST(AtomicWakerTest, ThreadStress) {
+  AtomicWaker waker;
+  std::vector<std::thread> threads;
+  std::atomic<bool> done{false};
+  std::atomic<int> wakeups{0};
+  std::atomic<int> drops{0};
+  std::atomic<int> armed{0};
+  std::atomic<int> not_armed{0};
+
+  threads.reserve(90);
+  for (int i = 0; i < 30; i++) {
+    threads.emplace_back([&] {
+      while (!done.load(std::memory_order_relaxed)) {
+        waker.Wakeup();
+      }
+    });
+  }
+  for (int i = 0; i < 30; i++) {
+    threads.emplace_back([&] {
+      while (!done.load(std::memory_order_relaxed)) {
+        class TestWakeable final : public Wakeable {
+         public:
+          TestWakeable(std::atomic<int>* wakeups, std::atomic<int>* drops)
+              : wakeups_(wakeups), drops_(drops) {}
+          void Wakeup() override {
+            wakeups_->fetch_add(1, std::memory_order_relaxed);
+            delete this;
+          }
+          void Drop() override {
+            drops_->fetch_add(1, std::memory_order_relaxed);
+            delete this;
+          }
+
+         private:
+          std::atomic<int>* const wakeups_;
+          std::atomic<int>* const drops_;
+        };
+        waker.Set(Waker(new TestWakeable(&wakeups, &drops)));
+      }
+    });
+  }
+  for (int i = 0; i < 30; i++) {
+    threads.emplace_back([&] {
+      while (!done.load(std::memory_order_relaxed)) {
+        (waker.Armed() ? &armed : &not_armed)
+            ->fetch_add(1, std::memory_order_relaxed);
+      }
+    });
+  }
+
+  do {
+    std::this_thread::sleep_for(std::chrono::seconds(1));
```

Comment:
If this has to loop and sleep for more than 1s, something is probably broken, right?

I think this is fine, but it's interesting to consider that from run to run, this may exercise the AtomicWaker a few times, or a few thousand times on the extremes of threading behavior under machine load and whatnot.

Reply:
yeah, most runs I expect one iteration here... if it takes longer it's not *bad* it's just weird CI load... if it takes forever (i.e. times out) it's broken.

Reply:
Since this test could sleep ~indefinitely under a broken implementation, how about setting an overall deadline at 5s then? If it is broken, I think the usual timeout for medium sized tests is 5 minutes.

Reply:
I'm gonna leave it as is: hesitant to add another source of potential flakiness if I get the overall time wrong this week.